purchase_last_price_info/models/product.py

```diff
@@ -0,0 +1,71 @@
+
+from odoo import api, fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    last_purchase_price = fields.Float(
+        string='Last Purchase Price')
+    last_purchase_date = fields.Date(
+        string='Last Purchase Date')
+    last_supplier_id = fields.Many2one(
+        comodel_name='res.partner', string='Last Supplier')
+
+    @api.multi
+    def set_product_last_purchase(self, order_id=False):
+        """ Get last purchase price, last purchase date and last supplier """
+        PurchaseOrderLine = self.env['purchase.order.line']
+        if not self.check_access_rights('write', raise_exception=False):
+            return
+        for product in self:
+            date_order = False
+            price_unit_uom = 0.0
+            last_supplier = False
+
+            # Check if Order ID was passed, to speed up the search
+            if order_id:
+                lines = PurchaseOrderLine.search([
+                    ('order_id', '=', order_id),
+                    ('product_id', '=', product.id)], limit=1)
+            else:
+                lines = PurchaseOrderLine.search(
+                    [('product_id', '=', product.id),
+                     ('state', 'in', ['purchase', 'done'])]).sorted(
+                    key=lambda l: l.order_id.date_order, reverse=True)
```

Comment:
why not limit to one here?

Reply:
I'm not sure it can be done since you are doing the search first, then sorting with the method sorted(). If you limit to 1 the search you may not get the most recent PO.

Reply:
You could try to use order='date_order desc' on search

Reply:
Well, at the end these are parameters that translate into an sql query (something like `SELECT * FROM purchase_order_line WHERE product_id = '3' AND state IN ('purchase', 'done') SORT BY date DESC`), so I'd say it's going to do it well at once

Reply:
The problem here is that you are searching in the purchase order line table but using the purchase order date to sort.